src/main/mondrian/olap/fun/MemberExtractingVisitor.java

```diff
@@ -0,0 +1,144 @@
+/*
+// This software is subject to the terms of the Eclipse Public License v1.0
+// Agreement, available at the following URL:
+// http://www.eclipse.org/legal/epl-v10.html.
+// You must accept the terms of that agreement to use this software.
+//
+// Copyright (c) 2002-2014 Pentaho Corporation..  All rights reserved.
+*/
+package mondrian.olap.fun;
+
+import mondrian.mdx.*;
+import mondrian.olap.*;
+import mondrian.olap.type.Type;
+
+import java.util.*;
+
+/**
+ * Visitor which collects any non-measure base members encountered while
+ * traversing an expression.
+ *
+ * This Visitor is used by the native set classes as well as the crossjoin
+ * optimizer (CrossjoinFunDef.nonEmptyList) to identify members within
+ * an expression which may conflict with members used as a constraint.
+ *
+ * If the boolean mapToAllMember is true, then any occurrences of
+ * a Dimension/Hierarchy/Level will result in the corresponding [All] member
+ * being added to the collection.  Likewise if a specific member is
+ * visited, the [All] member of it's corresponding hierarchy will be
+ * added.
+ *
+ * The mapToAllMember behavior will be used for any subexpression under
+ * one of the functions in the blacklist collection below.
+ */
+public class MemberExtractingVisitor extends MdxVisitorImpl {
+
+    private final Set<Member> memberSet;
+    private final ResolvedFunCallFinder finder;
+    private final Set<Member> activeMembers = new HashSet<Member>();
+    private final ResolvedFunCall call;
+    private final boolean mapToAllMember;
+
+    /**
+     * This list of functions are "blacklisted" because
+     * occurrence of a member/dim/level/hierarchy) within
+     * one of these expressions would result in a set of members
+     * that cannot be determined from the expression itself.
+     */
+    private static final String[] unsafeFuncNames = new String[] {
```

Comment:
I think you're right- it should be.  It would be worth a full review of the function list before merging.  There may be others omitted.